### PR TITLE
udev: extend timeout to prevent kill worker

### LINKED ISSUE #30436: Udevd worker process does not restore inotify watches when a child script times out and subsequently the worker is killed due to timeout.
### systemd version the issue has been seen with

Latest

### Used distribution

OL9

### Linux kernel version used

_No response_

### CPU architectures issue was seen on

x86_64

### Component

systemd, systemd-udevd

### Expected behaviour you didn't see

When a script (called using PROGRAM or RUN in a udev rule) times out (default timeout is 180 seconds, so this means the script takes more than 180 seconds to complete), udevd kills the worker process which spawned the script. When the worker gets killed like this through SIGKILL (because by default SIGKILL is used to kill the worker), it does not complete its remaining workflows as it exits immediately. If there was an inotify watch associated with this device, that would be suspended while processing the rule. That should get restored when the worker exits gracefully - but when the worker gets killed due to timeout, it does not get a chance to restore the watch. This means subsequent changes/writes do not trigger a change event for the device.

### Unexpected behaviour you saw

After a script timeout, further writes to a device (which has a inotify rule), do not trigger a udev event anymore.when the worker gets killed due to timeout, it does not get a chance to restore the watch. This means subsequent changes/writes do not trigger a change event for the device.

### Steps to reproduce the problem

The timeout and loss of inotify watch by udevd can be simulated by using a loop device (by default, these devices are watched using inotify, by udevd). I used a small C program to open the loop device in RDWR, which will trigger a inotify event. Upon receiving the event, udevd will synthesize a change uevent, which in turn will cause my custom script to get executed.

1) Rules file for loop device:-

cat /etc/udev/rules.d/90-my-loopdevice.rules
KERNEL=="loop1",  ACTION=="change",  RUN+="/root/script.sh"

2) Script to cause timeout while processing change event for loop device:-

cat /root/script.sh
#!/bin/sh
logger "script.sh - start"
sleep 185
logger "script.sh - end"

Steps to reproduce:-
1. Create a loop device, it automatically starts getting watched
2. Generate a open+close event for the device using the test program, which will trigger udev rule and will cause timeout
while processing the rule
3.  Remove the sleep from script, generate a open+close event for the device using the test program, verify that change event does not get processed by udev any more.

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #30436.
Replaces #30516.

### CODE REVIEW

**Comment by sushmbha** [comment]
Hi @yuwata , any idea when can commit [3c37d29](https://github.com/systemd/systemd/pull/30532/commits/3c37d291d18c70a185226d6985818a5a00593717) be merged into systemd:main ? I am hitting this issue and it would be helpful to have the fix merged so I can backport it.

**Comment by yuwata** [comment]
@sushmbha But, as you know, most of us are in holiday season...

File: src/udev/udev-spawn.c [hunk 1]
```diff
@@ -336,15 +337,26 @@ void udev_event_execute_run(UdevEvent *event, usec_t timeout_usec, int timeout_s
...
                                 log_device_debug_errno(event->dev, r, "Failed to run built-in command \"%s\", ignoring: %m", command);
                 } else {
-                        if (event->exec_delay_usec > 0) {
+                        if (event->worker && event->worker->exec_delay_usec > 0) {
+                                usec_t timeout_usec = event->worker ? event->worker->timeout_usec : DEFAULT_WORKER_TIMEOUT_USEC;
+                                usec_t now_usec = now(CLOCK_MONOTONIC);
+                                usec_t age_usec = usec_sub_unsigned(now_usec, event->birth_usec);
+
+                                if (event->worker->exec_delay_usec >= usec_sub_unsigned(timeout_usec, age_usec)) {
+                                        log_device_warning_errno(event->dev, r,
```
Comment (mrc0mmand) [inline src/udev/udev-spawn.c]: Shouldn't this be just `log_device_warning()`? Since `r` is used uninitialized here.

Comment (yuwata) [reply src/udev/udev-spawn.c]: Thanks. Fixed.

**Comment by yuwata** [comment]
Now this is ready for review. PTAL.

**Review by bluca — APPROVED** [review]

**Comment by yuwata** [comment]
jammy-ppc64el is failed in TEST-75-RESOLVED, unrelated.

**Pull request merged**